### PR TITLE
Support linux boot protocol for kexec.

### LINKED ISSUE #65: Support linux boot protocol for kexec
The live-bootstrap project boots Linux from Fiwix which requires support for [the linux boot protocol](https://www.kernel.org/doc/html/v5.6/x86/boot.html).

The implementation in the forthcoming PR is similar to the multiboot1 solution:
* The `kexec_protocol` with value `linux` must be specified at the kernel command line.
* The `kexec_size` parameter is used to configure a ram drive which holds the kernel (and initrd ram drive).
* The kexec is performed when the kernel halts.

There are several design points that should be explained.

The kexec process for live-bootstrap uses both a linux kernel and an initial ram file system.
These files are built under Fiwix and then must be passed to the Fiwix kernel for kexec.
Similar to multiboot, the files are copied to a ram drive device. However, since there are two files the process is a little more complicated as explained later.

Assuming you boot Fiwix from a hard drive (and so there is no initrd ram drive device) then the kexec ram drive will be /dev/ram0.

--

By the way...

Note that live-bootstrap boots with a very large initrd ram drive, which is configured with the Fiwix kernel parameter `ramdisksize=1179648` (approx. 1.2G). This is a problem because Fiwix compiles by default with one (additional) "general purpose" ram drive which uses the same parameter for sizing. But there is not enough memory for that.

Therefore, live-bootstrap compiles Fiwix with:
```
#define RAMDISK_DRIVES  0
```

In order to configure this setting, `RAMDISK_DRIVES` has been moved to include/fiwix/config.h.

--

A major way that linux boot differs from multiboot is the way that the size of the kernel and ram drive are conveyed to the Fiwix kernel for kexec. This is done by writing the length of the linux kernel in bytes to the ram drive as a 32-bit little-endian number followed by the length of the ram drive as another 32-bit little endian number. Then the kernel and ram drive are written to the
ram drive, sequentially, starting right after the lengths.

I am open to feedback on this design. An alternative is to configure multiple ram drives but that would require additional kexec kernel parameters. Moreover, the size of the linux kernel and initramfs must be specified exactly in the linux boot protocol but those sizes will not be known until after Fiwix is booted so we cannot use boot parameters to tell Fiwix those sizes. So, those sizes need to be conveyed to the Fiwix kernel somehow as part of initiating the kexec. The design I came up with allows using the same kexec kernel parameters that multiboot uses.

The Linux boot parameters takes more than one page so the size of the boot memory was increased to two pages and the location adjusted accordingly.

Note that the disabling IRQs does not work for booting Linux so I moved that code to after the kexec code.

Note that some of the code to prepare the linux boot parameters was adapted from the [Limine project](https://github.com/limine-bootloader/limine) and so the copyright notices reflect that.

--

Test files
---------

The following files are used to aid in testing:

```
linux-4.9.10
kexec-linux.c
initramfs.cpio
```

Unfortunately I could not upload initramfs.cpio.gz because at 133MB it was too large for github. I'm not sure how to send it to you but you can email me at grick23 at gmail.com and perhaps we can arrange something. Or you can follow the following instructions to generate the file yourself. In the mean time it is possible to test kexec without a initramfs as explained later.

Note, for future reference, the following section describes how these test files were produced.

Producing test files
--------------------

Clone and get ready to run the live-bootstrap project:

```
git clone https://github.com/fosslinux/live-bootstrap
cd live-bootstrap
```

In the file steps/jump/linux.sh replace this line:
```
    kexec-linux "/dev/ram1" "/boot/linux-4.9.10" "!$(command -v gen_init_cpio) /initramfs.list"
```

With this:

```
    $(command -v gen_init_cpio) /initramfs.list > /initramfs.cpio || true
    sync; read line
    kexec-linux "/dev/ram1" "/boot/linux-4.9.10" "!$(command -v gen_init_cpio) /initramfs.list"
```

Start live-bootstrap:
```
./rootfs.py --qemu
```
After 60 minutes or so live-bootstrap will hit the `read line` command and stop for input.
You will see the following on the screen:
```
linux-4.9.10: install to fakeroot.
linux-4.9.10: postprocess binaries.
linux-4.9.10: creating package.
linux-4.9.10: cleaning up.
linux-4.9.10: installing package.
linux-4.9.10: build successful
Unrecognized nod format '/dev/sda 600 0 0 b Hr Lr' line 7071
Unrecognized nod format '/dev/sda1 600 0 0 b Hr Lr' line 7072
Unrecognized nod format '/dev/sda2 600 0 0 b Hr Lr' line 7073
Unrecognized nod format '/dev/sda3 600 0 0 b Hr Lr' line 7074
Unrecognized nod format '/dev/sdb 600 0 0 b Hr Lr' line 7075
Unrecognized nod format '/dev/sdb1 600 0 0 b Hr Lr' line 7076
Unrecognized nod format '/dev/sdb2 600 0 0 b Hr Lr' line 7077
Unrecognized nod format '/dev/sdc 600 0 0 b Hr Lr' line 7078
Unrecognized nod format '/dev/sdc1 600 0 0 b Hr Lr' line 7079
Unrecognized nod format '/dev/sdc2 600 0 0 b Hr Lr' line 7080
Unrecognized nod format '/dev/sdc3 600 0 0 b Hr Lr' line 7081
```

At this point, you must go to the qemu console by pressing ctrl-a followed by c.
At the qemu prompt type the following commands, which will copy the ram disk to a file:

```
QEMU 4.2.1 monitor - type 'help' for more information
(qemu) pmemsave 0x001c6000 0x48000000 initrd.suspend
(qemu) quit
```

Now you can mount the ram disk and extract the linux kernel image, the ram drive cpio file used by linux, and the kexec-linux.c launching code:

```
cd ~/live-bootstrap
mkdir mnt
sudo mount -t ext2 initrd.suspend mnt
cp mnt/boot/linux-4.9.10 ~/
cp mnt/initramfs.cpio ~/
sudo umount mnt
cp steps/kexec-linux-1.0.0/files/kexec-linux.c ~/
```

Note that for testing on FiwixOS you'll need to change the following line in kexec-linux.c from this:
```
        reboot(RB_HALT_SYSTEM);
```
to this:
```
        system("halt");
```

The kexec-linux.c program I am providing here is also different in that it does not require providing an initial ram drive image.

--

Testing kexec linux
-------------------

Once you have the test files, you can try to kexec Linux from Fiwix.

You'll need to build the Fiwix kernel with the CONFIG_KEXEC parameter.

```
echo "#define CONFIG_KEXEC" > include/fiwix/custom_config.h
make clean
make CONFFLAGS="-DCUSTOM_CONFIG_H"
```

You'll need to copy the test files to the Fiwix hard drive.

Note: I'm using the same FiwixOS-3.2-i386.raw downloaded from Fiwix.org except that the S0 terminal has been uncommented near the bottom of /etc/inittab to enable /dev/ttyS0 so text / serial console works with qemu.

BEWARE: RUNNING THIS TEST WILL REFORMAT THE HARD DRIVE!
BEWARE: RUNNING THIS TEST WILL REFORMAT THE HARD DRIVE!
BEWARE: RUNNING THIS TEST WILL REFORMAT THE HARD DRIVE!

You'll need to use the correct qemu command line to launch Fiwix:

```
qemu-system-i386 \
    -enable-kvm \
    -nographic \
    -machine pc \
    -cpu 486 \
    -drive file=FiwixOS-3.2-i386.raw,format=raw,if=ide,cache=writeback,index=0 \
    -nic user,ipv6=off,model=e1000 \
    -kernel fiwix \
    -append "console=/dev/ttyS0 root=/dev/hda2 kexec_proto=linux kexec_size=280000 kexec_cmdline=\"init=/init console=ttyS0\"" \
    -m size=4G
```

Login to Fiwix as root/root and run:

```
cc -m32 kexec-linux.c -o kexec-linux
./kexec-linux /dev/ram0 ./linux-4.9.10 ./initramfs.cpio
# After about 30 seconds linux should boot and start running gcc commands
```

Note that an initramfs image is optional. If not used, a size of zero should be written to the ram drive.
I used this to boot Linux directly with a hard drive root using the FiwixOS 3.2 image.
Yes, Linux can boot FiwixOS!

The steps required for this are to change the qemu append parameter:
```
    -append "console=/dev/ttyS0 root=/dev/hda2 kexec_proto=linux kexec_size=280000 kexec_cmdline=\"root=/dev/sda2 init=/sbin/init console=ttyS0\"" \
```

After booting Fiwix, change device from hda to sda:
```
# sed -i 's/hda/sda/g' /etc/rc.d/rc.sysinit
```

Then run kexec-linux without an initramfs file name:
```
# ./kexec-linux /dev/ram0 ./linux-4.9.10
```

[kexec-linux.c.gz](https://github.com/mikaku/Fiwix/files/13833407/kexec-linux.c.gz)
[linux-4.9.10.gz](https://github.com/mikaku/Fiwix/files/13833412/linux-4.9.10.gz)

### PR DESCRIPTION
**Pull request by rick-masters**
Resolves #65

### CODE REVIEW

File: kernel/kexec.c [hunk 1]
```diff
@@ -1,7 +1,9 @@
 /*
  * fiwix/kernel/kexec.c
  *
- * Copyright 2023, Jordi Sanfeliu. All rights reserved.
```
Comment (mikaku) [inline kernel/kexec.c:4]: What is the purpose to change from 2023 to 2024 this copyright?

File: kernel/main.c [hunk 2]
```diff
@@ -144,12 +144,6 @@ void stop_kernel(void)
 	struct proc *p, *next;
 	int n;
 
-	/* stop and disable all interrupts! */
-	CLI();
-	for(n = 0; n < NR_IRQS; n++) {
-		disable_irq(n);
-	}
-
```
Comment (mikaku) [inline kernel/main.c:152]: Just wondering what was the problem with Linux to have enabled interrupts. Does this is part of the Linux boot specification?

File: kernel/main.c [hunk 3]
```diff
@@ -176,6 +173,12 @@ void stop_kernel(void)
 	printk("** Press Any Key to Reboot **\n");
 	any_key_to_reboot = 1;
 
+	/* stop and disable all interrupts! */
+	CLI();
+	for(n = 0; n < NR_IRQS; n++) {
+		disable_irq(n);
+	}
+
```
Comment (mikaku) [inline kernel/main.c:181]: I think that interrupts should be disabled **before** the last kernel messages.

File: kernel/kexec.c (hunk 1 above)
Comment (rick-masters) [reply kernel/kexec.c:4]: I was just updating all copyrights in this file to the current year but I can revert that back to 2023 if you prefer. Please let me know.

Comment (mikaku) [reply kernel/kexec.c:4]: Yes, I think we must keep, at least, the original year as this will help to know better when this file was created.

Comment (rick-masters) [reply kernel/kexec.c:4]: OK I will change that back to 2023.

Comment (mikaku) [reply kernel/kexec.c:4]: Thanks.

I still have pending to make some adjustments in the code, related to the coding style.

File: kernel/main.c (hunk 2 above)
Comment (rick-masters) [reply kernel/main.c:152]: I just checked and the specifications says to disable interrupts with `cli` before booting Linux.

I move the disable back to where it was and Linux appears to boot OK so I am not sure what the problem was before. I am running a full test under live-bootstrap (which takes over an hour) to see if this code can remain where it was. However, please see the comment below for another reason I moved the code.

File: kernel/main.c (hunk 3 above)
Comment (rick-masters) [reply kernel/main.c:181]: > I think that interrupts should be disabled **before** the last kernel messages.

Could you elaborate on why you would prefer that? My experience has been that if you are using a serial port console, this prevents the last messages from being displayed and pressing a key does nothing. This is the output I see with a serial port:

```
Welcome to FiwixOS 3.2 (ttyS0)
Kernel Release 1.5.0

(none) login: root
Password: 

Welcome to FiwixOS 3.2 for the i386 architecture.

    - Read the /root/README file to know about all the files installed.
    - Refer to the Installation CD-ROM for the source code and the license of
      every single package installed in this system.

[(root) ~]# shutdown -h 0

Broadcast message from root (ttyS0) (Fri Jan  5 17:23:47 2024):

The system is going down for system halt NOW!
INIT: Switching to runlevel: 0
INIT: Sending processes the TERM signal
INIT: Sending processes the KILL signal
Stopping atd:                                              [  OK  ]
Stopping crond:                                            [  OK  ]
WARNING: root 2 with count 10 (on dev 3,2)
WARNING: inode 2 with count 10 (on dev 3,2)
umount: /: Device or resource busy
Remounting remaining filesystems (if any) readonly
Halting the system...
```

When I move the IRQ disabling *after* printing the last kernel messages then I do see the kernel messages. It appears that the serial port I/O requires enabled interrupts. However, pressing a key does not reboot the kernel when using a serial port so it didn't fully solve the problem but I thought it was better.

Comment (mikaku) [reply kernel/main.c:181]: Yes, you're right. The last kernel messages does not appear on a serial console, but the cause is not really this although disabling interrupts **after** the messages seems fixes it.

The problem is in the serial console driver that is not managing correctly the serial port when the messages come from the kernel. You can test it using the magic SysRq key. If you press `Alt+SysRq+t' to display the current list of processes, you'll see that the list is incomplete (depending on the number of processes) on a serial console. I'm aware of this problem and I plan to fix it during the 1.5.0 version. Remember, I advice you some time ago, to increase the value of `NR_CB_QUEUE` in `include/fiwix/tty.h` if you wanted to see all that process list.

Meanwhile, leave the code as it is now. I mean, to disable interrupts **after** the kernel messages, so you'll notice if the kernel halted successfully, and so you'll have a better debugging experience.

Once I fixed the bug, I'll change the code to be **before** of the last kernel messages.

Comment (mikaku) [reply kernel/main.c:181]: By the way, I recommend you to use the latest FiwixOS 3.3 version.

**Comment by rick-masters** [comment]
If there is anything else I need to work on please let me know. I know you are working on coding / style changes.

**Comment by mikaku** [comment]
> If there is anything else I need to work on please let me know. I know you are working on coding / style changes.

All done.

Please, rebase it and let me know if all is working fine.
Then I'll merge it.

**Comment by rick-masters** [comment]
> Please, rebase it and let me know if all is working fine. Then I'll merge it.

Looks ready to me. Thanks!

**Comment by mikaku** [comment]
Did you change the copyright year?

**Comment by rick-masters** [comment]
> Did you change the copyright year?

Oh yeah, just pushed that.

**Comment by mikaku** [comment]
Thank you.

**Pull request merged**